x-pack/plugins/alerting/server/task_runner/task_runner.ts

```diff
@@ -589,6 +596,62 @@ export class TaskRunner<
   }
 }
 
+interface TrackAlertDurationsParams<
+  InstanceState extends AlertInstanceState,
+  InstanceContext extends AlertInstanceContext
+> {
+  originalAlerts: Dictionary<AlertInstance<InstanceState, InstanceContext>>;
+  currentAlerts: Dictionary<AlertInstance<InstanceState, InstanceContext>>;
+  recoveredAlerts: Dictionary<AlertInstance<InstanceState, InstanceContext>>;
+}
+
+function trackAlertDurations<
+  InstanceState extends AlertInstanceState,
+  InstanceContext extends AlertInstanceContext
+>(params: TrackAlertDurationsParams<InstanceState, InstanceContext>) {
+  const currentTime = new Date().toISOString();
+  const { currentAlerts, originalAlerts, recoveredAlerts } = params;
+  const originalAlertIds = Object.keys(originalAlerts);
+  const currentAlertIds = Object.keys(currentAlerts);
+  const recoveredAlertIds = Object.keys(recoveredAlerts);
+  const newAlertIds = without(currentAlertIds, ...originalAlertIds);
+
+  // Inject start time into instance state of new instances
+  for (const id of newAlertIds) {
+    const state = currentAlerts[id].getState();
+    currentAlerts[id].replaceState({ ...state, start: currentTime, uuid: uuidv4() });
```

Comment:
side thought, for @YulNaumenko - does this new UUID value for the alert id belong in one of the new `rule.` fields we're writing?

Reply:
Yeah, makes sense to me. Need to check with @elastic/security-detections-response

Reply:
I don't think it makes sense to write `alert` information into the `rule` object?

Reply:
Good philosophical question.  An alert is rule-related, so it seems fine to me to.  As an example, [the `process.` fields have a `thread.id`](https://www.elastic.co/guide/en/ecs/current/ecs-process.html#field-process-thread-id) - kinda seems like the same kind of relationship.  We might be able to get some historic context from the RFC that introduced this.

I guess the RAC field for this value is `kibana.rac.alert.uuid` right now?  Per previous comment ^^^, I'm hesitant to add `kibana.rac` fields right now, especially if it's something that _should_ be in `rule.` (or some other existing field).  Maybe we need to get a new ECS RFC together to add such a field to `rule.`?  Let's bring this up with the RAC alerts-as-data folk ...